File: lv1_module2/ros2_ws/src/turtle_cpp/src/distance_publisher.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <cmath>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/pose.hpp"
#include "std_msgs/msg/float32.hpp"
#include "rcl_interfaces/msg/set_parameters_result.hpp"

using namespace std::chrono_literals;

class DistancePublisher : public rclcpp::Node {
public:
    DistancePublisher() : Node("turtle_distance_publisher"), current_rate_(10.0) {
        // 1. 파라미터 선언 및 기본값 초기화 (10.0 Hz)
        this->declare_parameter<double>("publish_rate", 10.0);
        this->get_parameter("publish_rate", current_rate_);

        // 2. Publisher 및 Subscriber 설정
        publisher_ = this->create_publisher<std_msgs::msg::Float32>("/turtle_distance", 10);
        subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtle1/pose", 10, std::bind(&DistancePublisher::pose_callback, this, std::placeholders::_1));

        // 3. 타이머 생성 (10Hz)
        auto period = std::chrono::duration<double>(1.0 / current_rate_);
        timer_ = this->create_wall_timer(period, std::bind(&DistancePublisher::timer_callback, this));

        // 4. 파라미터 변경 동적 콜백 등록
        param_callback_handle_ = this->add_on_set_parameters_callback(
            std::bind(&DistancePublisher::parameter_callback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "C++ Distance Publisher Node Initialized.");
    }

private:
    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg) {
        current_pose_ = msg;
    }

    void timer_callback() {
        if (!current_pose_) {
            return;
        }

        // 원점(0,0)으로부터 직선 거리 계산 (hypot)
        double distance = std::hypot(current_pose_->x, current_pose_->y);

        auto msg = std_msgs::msg::Float32();
        msg.data = static_cast<float>(distance);
        publisher_->publish(msg);
    }

    rcl_interfaces::msg::SetParametersResult parameter_callback(const std::vector<rclcpp::Parameter> &parameters) {
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;

        for (const auto &param : parameters) {
            if (param.get_name() == "publish_rate" && param.get_type() == rclcpp::ParameterType::PARAMETER_DOUBLE) {
                double new_rate = param.as_double();
                if (new_rate <= 0.0) {
                    result.successful = false;
                    result.reason = "Publish rate must be positive.";
                    return result;
                }

                if (new_rate != current_rate_) {
                    RCLCPP_INFO(this->get_logger(), "Publish rate changed: %.1f -> %.1f Hz", current_rate_, new_rate);
                    current_rate_ = new_rate;
                    timer_->cancel();
                    auto period = std::chrono::duration<double>(1.0 / current_rate_);
                    timer_ = this->create_wall_timer(period, std::bind(&DistancePublisher::timer_callback, this));
                }
            }
        }
        return result;
    }

    double current_rate_;
    turtlesim::msg::Pose::SharedPtr current_pose_{nullptr};
    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DistancePublisher>();
    try {
        rclcpp::spin(node);
    } catch (const std::exception &e) {
        (void)e;
    }
    rclcpp::shutdown();
    return 0;
}

File: lv1_module2/ros2_ws/src/turtle_py/turtle_py/distance_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import math

from turtlesim.msg import Pose
from std_msgs.msg import Float32
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter

class DistancePublisher(Node):
    def __init__(self):
        super().__init__('turtle_distance_publisher')
        
        # publish_rate 파라미터 선언 및 초기화 (기본값: 10.0 Hz)
        self.declare_parameter('publish_rate', 10.0)
        self.current_rate = self.get_parameter('publish_rate').get_parameter_value().double_value
        
        # 최신 위치 보관 변수
        self.current_pose = None
        
        # Publisher & Subscriber 구성
        self.publisher_ = self.create_publisher(Float32, '/turtle_distance', 10)
        self.subscription = self.create_subscription(
            Pose,
            '/turtle1/pose',
            self.pose_callback,
            10
        )
        
        # 10Hz 기본 타이머 생성
        timer_period = 1.0 / self.current_rate
        self.timer = self.create_timer(timer_period, self.timer_callback)
        
        # 동적 파라미터 변경 콜백 등록
        self.add_on_set_parameters_callback(self.parameter_callback)
        self.get_logger().info("Distance Publisher Node Initialized.")

    def pose_callback(self, msg):
        # 최신 포즈를 보관만 수행
        self.current_pose = msg

    def timer_callback(self):
        if self.current_pose is None:
            return
            
        # 원점(0,0)으로부터 현재 거북이 위치(x,y)까지의 직선 거리 계산
        distance = math.hypot(self.current_pose.x, self.current_pose.y)
        
        # 토픽 발행
        msg = Float32()
        msg.data = distance
        self.publisher_.publish(msg)

    def parameter_callback(self, params):
        for param in params:
            if param.name == 'publish_rate' and param.type_ == Parameter.Type.DOUBLE:
                new_rate = param.value
                if new_rate <= 0.0:
                    return SetParametersResult(successful=False, reason="Publish rate must be positive.")
                
                if new_rate != self.current_rate:
                    self.get_logger().info(f"Publish rate changed: {self.current_rate} -> {new_rate} Hz")
                    self.current_rate = new_rate
                    self.destroy_timer(self.timer)
                    self.timer = self.create_timer(1.0 / self.current_rate, self.timer_callback)
                    
        return SetParametersResult(successful=True)

def main(args=None):
    rclpy.init(args=args)
    node = DistancePublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()

